### DISCUSSION TITLE
Preview panel

### DISCUSSION THREAD

**Discussion by JordanMarr** [fsprojects/Avalonia.FuncUI#284] [Ideas]
I wonder what it would take to get the Avalonia preview panel to work with FuncUI. I'm assuming it must take in an xml string from the .axaml file. If it is possible to serialize a FuncUI hierarchy to XML, then maybe the Avalonia preview panel could be modified to natively work with FuncUI.

**Comment by JaggerJo** [comment]
Having the previewer work would be awesome, I have no idea what would be necessary tho.

Maybe @ForNeVeR could point us in the right direction here?

**Reply by ForNeVeR** [reply]
The current Avalonia Previewer works because XAML reload is possible. The same can't be said for the F# code, AFAIK.

But still, something's possible to do here. Maybe start from "auto-build + reload" approach?

Sorry, I am not so fluent in FuncUI, so something I say may be a bit off.

We'll need a way to identify a FuncUI component (for the IDE to say to the previewer "I want you to show a component `name`", and then some way (maybe a custom protocol command) for the previewer process to construct the aforementioned component.

The latter may very well reuse existing infrastructure in the Avalonia Previewer.

If these two basic requirements are met, then we'll be able to implement the preview in any IDE, using the existing Avalonia infrastructure. It won't work too fast (yet!), but it will be already something useful.

**Reply by ForNeVeR** [reply]
Auto-generating of a XAML file proposed by @JordanMarr is also an exciting idea. Avalonia Previewer already has a facility that accepts a XAML text and shows it in a preview window. Some small modifications to make it work without _the initial_ XAML file, and I guess something better (than my initial proposal) may come out?

**Reply by JaggerJo** [reply]
Hey @ForNeVeR 👋

I think one important difference between XAML preview and F# preview is that one F# file often contains more than one component. I basically have a bare bones version of what you suggested (auto-build + reload) running - see my comment below.

**Comment by JaggerJo** [comment]
@JordanMarr I guess we'd need to build something on top of the [Avalonia Designer Project](https://github.com/AvaloniaUI/Avalonia/tree/cbbc27f631fa7cb5faa880af6a80884b9372c0b9/src/Avalonia.DesignerSupport). 

We'd need to know the function (constrained by `args/unit ->  Component`) to preview.

**Comment by JaggerJo** [comment]
Worked on a prototype this evening. 

Pretty simple actually, the previewer app just watches a dll for changes, `dotnet watch` is then run form the application we want to see "live".


https://user-images.githubusercontent.com/13090415/227038831-39cb5448-72fa-413b-a0a5-de1f521df40d.mp4

(had to split the recoding)

https://user-images.githubusercontent.com/13090415/227039065-9a8d8e84-ea64-43b9-a046-5f0a3ee70619.mp4

**Reply by JaggerJo** [reply]
I think the main issue with this approach is that we actually load (and unload) the dll's for the app we are previewing. 

I'm not sure what would happen if the previewer would use a different version of Avalonia to load the DLL.

Otherwise this works surprisingly well.

**Reply by JordanMarr** [reply]
That's exciting!

**Reply by JaggerJo** [reply]
Ahh, forgot to mention / might not be clear - the components in the previewer are "real".

![CleanShot 2023-03-22 at 22 26 09](https://user-images.githubusercontent.com/13090415/227042330-bd139349-bcd8-42e6-9cbb-2f2baba639bd.gif)

**Reply by JordanMarr** [reply]
Yeah, this totally needs to be a thing.

**Reply by JaggerJo** [reply]
It's "hacked in an hour" but here is the code if you want to try it

 https://github.com/fsprojects/Avalonia.FuncUI/tree/live-view-experiments

**Reply by JordanMarr** [reply]
I tried it out. It does display the counter app! Very cool!
What are you doing to get the preview to refresh? Saving doesn't work, and building crashes.

I also tried it on `Examples.ChordParser`, but it was unable to load Elmish.dll.

**Reply by JaggerJo** [reply]
you need to 'dotnet watch build' the counter app

**Reply by edgarfgp** [reply]
Looks promising 😀. I wish we could preview iOS and android as well

**Comment by Numpsy** [comment]
Sounds interesting, I'll try to give it a go later

**Comment by fischgeek** [comment]
Looks great guys! Very excited for this!
